File: todo.py
from js import document, console
from pyodide import create_proxy


class TodoList:
    def __init__(self, parent_element):
        self.parent_element = parent_element
        self.current_id = 0
        self.__init__input()
        self.__init__button()
        self.__init__list()
        pass

    def __init__button(self):
        self.button_element = document.createElement("button")
        self.button_element.innerText = "add todo"
        self.parent_element.append(self.button_element)
        add_todo = create_proxy(self.add_todo)
        self.button_element.addEventListener("click", add_todo)
        pass

    def __init__input(self):
        self.input_element = document.createElement("input")
        self.parent_element.append(self.input_element)
        pass

    def __init__list(self):
        self.list_element = document.createElement("ul")
        self.parent_element.append(self.list_element)
        pass

    def add_todo(self, event):
        text = self.input_element.value
        new_item = Item(text, self.current_id)
        self.current_id = self.current_id + 1
        new_item.on_remove(self.remove_todo)
        self.list_element.append(new_item.item)
        pass

    def remove_todo(self, event):
        target_element = event.target.closest("li")
        self.list_element.removeChild(target_element)
        pass


class Item:
    def __init__(self, text, id):
        self.item = document.createElement("li")
        self.item.setAttribute("data-id", id)
        self.item.innerText = text
        self.remove_button = document.createElement("button")
        self.remove_button.innerText = "remove"
        self.item.append(self.remove_button)
        pass

    def on_remove(self, cb):
        on_remove_proxy = create_proxy(cb)
        self.remove_button.addEventListener("click", on_remove_proxy)
        pass


app = document.getElementById("app")

TodoList(app)

File: todo.js
class TodoList{
    constructor(parentElement){
        this.parentElement = parentElement;
        this.currentId = 0;
        this.initInput();
        this.initButton();
        this.inputList();
    }
    initButton(){
        this.buttonElement = document.createElement('button');
        this.buttonElement.innerText = "add todo";
        this.parentElement.append(this.buttonElement);
        this.buttonElement.addEventListener("click",this.addTodo.bind(this));
    }
    initInput(){
        this.inputElement = document.createElement('input');
        this.parentElement.append(this.inputElement);
    }
    inputList(){
        this.listElement = document.createElement('ul');
        this.parentElement.append(this.listElement);
    }
    addTodo(){
        const text = this.inputElement.value;
        const newItem = new Item(text,this.currentId);
        this.currentId = this.currentId + 1;
        newItem.onRemove(this.removeTodo.bind(this));
        this.listElement.append(newItem.item);
    }
    removeTodo(event){
        const targetElement = event.target.closest('li');
        this.listElement.removeChild(targetElement);
    }
}

class Item{
    constructor(text,id){
        this.item = document.createElement('li');
        this.item.setAttribute('data-id',id);
        this.item.innerText = text;
        this.removeButton = document.createElement('button');
        this.removeButton.innerText = 'remove';
        this.item.append(this.removeButton);
    }
    onRemove(cb){
        this.removeButton.addEventListener('click',cb);
    }
}
const app = document.getElementById('app');

new TodoList(app);